pkg/trimaran/targetloadpacking/targetloadpacking.go

```diff
@@ -0,0 +1,266 @@
+
+	var missingCPUUtilMillis int64 = 0
+	pl.eventHandler.RLock()
+	defer pl.eventHandler.RUnlock()
+	if _, ok := pl.eventHandler.ScheduledPodsCache[nodeName]; ok {
+		for _, info := range pl.eventHandler.ScheduledPodsCache[nodeName] {
+			// If the time stamp of the scheduled pod is outside fetched metrics window, or it is within metrics reporting interval seconds, we predict util.
+			// Note that the second condition doesn't guarantee metrics for that pod are not reported yet as the 0 <= t <= 2*metricsAgentReportingIntervalSeconds
+			// t = metricsAgentReportingIntervalSeconds is taken as average case and it doesn't hurt us much if we are
+			// counting metrics twice in case actual t is less than metricsAgentReportingIntervalSeconds
+			if info.Timestamp > metrics.Window.End || info.Timestamp <= metrics.Window.End &&
+				(metrics.Window.End-info.Timestamp) < metricsAgentReportingIntervalSeconds {
+				for _, container := range info.Pod.Spec.Containers {
+					missingCPUUtilMillis += PredictUtilisation(&container)
+				}
+				missingCPUUtilMillis += info.Pod.Spec.Overhead.Cpu().MilliValue()
+				klog.V(6).Infof("missing utilisation for pod %v : %v", info.Pod.Name, missingCPUUtilMillis)
+			}
+		}
+		klog.V(6).Infof("missing utilisation for node %v : %v", nodeName, missingCPUUtilMillis)
+	}
+
+	var predictedCPUUsage float64
+	if nodeCPUCapMillis != 0 {
+		predictedCPUUsage = 100 * (nodeCPUUtilMillis + float64(curPodCPUUsage) + float64(missingCPUUtilMillis)) / nodeCPUCapMillis
+	}
+	if predictedCPUUsage > hostCPUThresholdPercent {
+		if predictedCPUUsage > 100 {
+			return framework.MinNodeScore, framework.NewStatus(framework.Success, "")
+		}
+		penalisedScore := int64(math.Round(50 * (100 - predictedCPUUsage) / (100 - hostCPUThresholdPercent)))
+		klog.V(6).Infof("penalised score for host %v: %v", nodeName, penalisedScore)
+		return penalisedScore, framework.NewStatus(framework.Success, "")
+	}
+
+	score := int64(math.Round((100-hostCPUThresholdPercent)*predictedCPUUsage/hostCPUThresholdPercent + hostCPUThresholdPercent))
+	klog.V(6).Infof("score for host %v: %v", nodeName, score)
+	return score, framework.NewStatus(framework.Success, "")
+}
+
+func (pl *TargetLoadPacking) ScoreExtensions() framework.ScoreExtensions {
+	return pl
+}
+
+func (pl *TargetLoadPacking) NormalizeScore(ctx context.Context, state *framework.CycleState, pod *v1.Pod, scores framework.NodeScoreList) *framework.Status {
+	return nil
+}
+
+func updateArguments(obj runtime.Object) error {
+	args, err := getArgs(obj)
+	if err != nil {
+		return err
+	}
+	if args.DefaultCPURequests != nil && !args.DefaultCPURequests.Cpu().IsZero() {
+		requestsMilliCores = args.DefaultCPURequests.Cpu().MilliValue()
+	}
+	if args.TargetCPUUtilization != nil && *args.TargetCPUUtilization != 0 {
+		hostCPUThresholdPercent = *args.TargetCPUUtilization
+	}
+	if args.WatcherAddress != nil && *args.WatcherAddress != "" {
```

Comment:
What if this is nil? IIUC this plugin won't work - it will constantly report "new watcher request failed:".

Reply:
It is assumed load watcher would be available on localhost. But since you mentioned the idea of `scheduler-plugins` is to have a single image with multiple plugins, I will handle this logic differently now with no default.